openedx/core/djangoapps/programs/utils.py

```diff
@@ -126,6 +126,43 @@ def engaged_programs(self):
 
         return programs
 
+    @cached_property
+    def failed_course_runs(self):
+        """
+        Determine which course runs have been failed by the user.
+
+        Returns:
+            list of strings, each representing a certificate course run key
+        """
+        course_run_certificates = certificate_api.get_certificates_for_user(self.user.username)
+        return [
+            unicode(certificate['course_key']) for certificate in course_run_certificates
+            if not certificate_api.is_passing_status(certificate['status'])
+        ]
+
+    def does_course_still_qualify_as_in_progress(self, now, enrolled_run_modes, course):
```

Comment:
(also if you were suggesting i don't modify the object in that method, if i combined it with in progress, i wouldn't be able to tell apart the courses that should go into in progress vs not_started) so there are two separate categories i need to be able to check for within in progress